build: prioritize large packages

dmnks opened the pull request:
Binary packages come in different sizes and so their build time can vary
greatly.  Dynamic scheduling, which we currently use for parallel
building, is a good strategy to combat such differences and load-balance
the available CPU cores.

That said, knowing that the build time of a package is proportional to
its size, we can reduce the overall time even further by cleverly
ordering the task queue.

As an example, consider a set of 5 packages, 4 of which take 1 unit of
time to build and one takes 4 units.  If we were to build these on a
dual-core system, one possible unit distribution would look like this:

```
            TIME --->
CPU 1       * * * * * *         # package 1, 3 and 5
CPU 2       * *                 # package 2 and 4
```

Now, compare that to a different distribution where the largest package
5 gets built early on:

```
            TIME --->
CPU 1       * * * *             # package 5
CPU 2       * * * *             # package 1, 2, 3 and 4
```

It's obvious that processing the largest packages first gives better
results when dealing with such a mix of small and large packages
(typically a regular package and its debuginfo counterpart,
respectively).

Now, with dynamic scheduling in OpenMP, we cannot directly control the
task queue; we can only generate the tasks and let the runtime system do
its work.  What we can do, however, is to provide a hint to the runtime
system for the desired ordering, using the "priority" clause.

So, in this commit, we use the clause to assign a priority value to each
build task based on the respective package size (the bigger the size,
the higher the priority), to help achieve an optimal execution order.

Indeed, in my testing, the priorities were followed to the letter (but
remember, that's not guaranteed by the specification).  Interestingly,
even without the use of priorities, simply generating the tasks in the
desired order resulted in the same execution order for me, but that's,
again, just an implementation detail.

Also note that OpenMP is allowed to stop the thread generating the tasks
at any time, and make it execute some of the tasks instead.  If the
chosen task happens to be a long-duration one, we might hit a starvation
scenario where the other threads have exhausted the task queue and
there's nobody to generate new tasks.  To counter that, this commit also
adds the "untied" clause which allows other threads to pick up where the
generating thread left off, and continue generating new tasks.

Resolves #1045.

dmnks commented:
A testing spec file that I used to benchmark this patch:

```
%global _smp_build_nthreads 2

Name:           foo
Version:        1.0
Release:        1%{?dist}
Summary:        testing package for parallel building

License:        GPL

%description

%package light1
Summary:        light subpackage 1
%description light1

%package light2
Summary:        light subpackage 2
%description light2

%package light3
Summary:        light subpackage 3
%description light3

%package light4
Summary:        light subpackage 4
%description light4

%package heavy1
Summary:        heavy subpackage 1
%description heavy1

%install
mkdir -p %{buildroot}/%{_libdir}
dd if=/dev/urandom of=%{buildroot}/%{_libdir}/libfoo_light1.so count=30 bs=10M
dd if=/dev/urandom of=%{buildroot}/%{_libdir}/libfoo_light2.so count=30 bs=10M
dd if=/dev/urandom of=%{buildroot}/%{_libdir}/libfoo_light3.so count=30 bs=10M
dd if=/dev/urandom of=%{buildroot}/%{_libdir}/libfoo_light4.so count=30 bs=10M
dd if=/dev/urandom of=%{buildroot}/%{_libdir}/libfoo_heavy1.so count=120 bs=10M

%files light1
%{_libdir}/libfoo_light1.so
%files light2
%{_libdir}/libfoo_light2.so
%files light3
%{_libdir}/libfoo_light3.so
%files light4
%{_libdir}/libfoo_light4.so
%files heavy1
%{_libdir}/libfoo_heavy1.so

%changelog
* Mon Mar  2 2020 Michal Domonkos <mdomonko@redhat.com>
- 
```

On my laptop, the total build time is `~1:35s` and `~1:20s` before and after the patch, respectively. Not a huge difference here, but thread utilization is be a bit better now and should help with the really-big packages like the kernel (haven't tried that yet).

pmatilai commented on build/pack.c:
```diff
@@ -726,16 +727,37 @@ static rpmRC packageBinary(rpmSpec spec, Package pkg, const char *cookie, int ch
+    Package pkg2 = *(Package *)p2;
+    uint64_t size1 = headerGetNumber(pkg1->header, RPMTAG_SIZE);
+    uint64_t size2 = headerGetNumber(pkg2->header, RPMTAG_SIZE);
+    return size1 - size2;
```
The difference between two 64bit integers can overflow the 32bit int return code, you'll want to normalize the return one way or the other.

pmatilai commented on build/pack.c:
```diff
@@ -726,16 +727,37 @@ static rpmRC packageBinary(rpmSpec spec, Package pkg, const char *cookie, int ch
+    tasks = xcalloc(i, sizeof(Package));
+    for (i = 0, pkg = spec->packages; pkg != NULL; i++, pkg = pkg->next)
+        tasks[i] = pkg;
+    qsort(tasks, i, sizeof(Package), compareBinaries);
```
Do use separate variables for indexing and things like total package count, (re)using an 'i' for multiple different purposes only makes the code harder to follow for now good reason. Using loop-local index variables is good way to ensure there are no side-effects from previous rounds, and also makes it obvious to the reader as well.

```
    int npkgs = 0;

    for (pkg = spec->packages; pkg != NULL; pkg = pkg->next)
        npkgs++;
    tasks = xcalloc(npkgs, sizeof(Package));
    for (int i = 0, pkg = spec->packages; pkg != NULL; i++, pkg = pkg->next)
        tasks[i] = pkg;
    qsort(tasks, npkgs, sizeof(Package), compareBinaries);
```

pmatilai commented:
FWIW, I can easily reproduce similar timings here. Using `rpmbuild -bb --noclean --short-circuit' to limit it to only the packaging phase, I get ~58s before the patch and ~44s after it. It's quite a nice win.

BTW bonus points for the graphs in the commit message, they illustrate the problem really nicely.

dmnks commented:
@pmatilai Fixed and rebased.
Note that I changed `RPMTAG_SIZE` to `RPMTAG_LONGSIZE` because I realized the former may not be set if the total size is greater than 32 bits (details [here](https://rpm.org/devel_doc/large_files.html)).

pmatilai commented:
Good catch on the size tag, totally managed to miss that.
Looks fine by me, just one thing for consideration: since we want them largest first, might as well sort them that way too so the second for-loop doesn't need to walk backwards, which looks a bit more special than it actually is. Not that it makes any practical difference so up to you.

dmnks commented:
Good point on the backwards loop; it really shouldn't scream "look ma, reverse loop for no obvious reason, go figure out yourself!".

Instead, I'll just make the reverse ordering explicit by simply inverting the `compareBinaries()` function, and push a fixup commit right away.
